File: min_stack.rb
class MinStack
  def initialize()
    @val = []
    @min = []
  end

  def push(x)
    @val.push x
    cur_min = @min.last
    cur_min = Float::INFINITY if cur_min.nil?
    if x < cur_min
      cur_min = x
    end
    @min.push cur_min
  end

  def pop()
    @min.pop
    @val.pop
  end

  def top()
    @val.last
  end

  def get_min()
    @min.last
  end
end

# Your MinStack object will be instantiated and called as such:
obj = MinStack.new()

obj.push(-2)
obj.push(0)
obj.push(-3)

puts obj.get_min
puts obj.pop
puts obj.top
puts obj.get_min

File: amazon/src/main/java/design/MinStack.java
package design;

import java.util.LinkedList;

public class MinStack {
  LinkedList<Integer> q, m;
  public MinStack() {
    q = new LinkedList<>();
    m = new LinkedList<>();
  }

  public void push(int val) {
    q.push(val);
    int curMin = val;
    if (!m.isEmpty()) {
      curMin = Math.min(curMin, m.peek());
    }
    m.push(curMin);
  }

  public void pop() {
    q.pop();
    m.pop();
  }

  public int top() {
    return q.peek();
  }

  public int getMin() {
    return m.peek();
  }
}
